server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -3097,8 +3097,9 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
 
         try {
             closeShards(corruptedShard);
-        } catch (RuntimeException e) {
+        } catch (AssertionError e) {
             // Ignored because corrupted shard can throw various exceptions on close
+            assertThat(e.getCause(), either(instanceOf(RuntimeException.class)).or(instanceOf(CorruptIndexException.class)));
```

Comment:
Aha ok I didn't find `MockDirectoryWrapper` before. `catch (AssertionError e)` is pretty ugly. How about this hack instead?

```
        // Close the directory under the shard first because it's probably a MockDirectoryWrapper which throws exceptions when corrupt
        try {
            ((FilterDirectory) corruptedShard.store().directory()).getDelegate().close();
        } catch (CorruptIndexException | RuntimeException e) {
            // ignored
        }

        closeShards(corruptedShard);
```

Reply:
++ that one looks much nicer :)